src/repr/src/row.rs

```diff
@@ -523,6 +574,44 @@ unsafe fn read_datum<'a>(data: &'a [u8], offset: &mut usize) -> Datum<'a> {
             let t = Timestamp::decode(read_byte_array(data, offset));
             Datum::MzTimestamp(t)
         }
+        Tag::Range => {
+            let flags = read_byte(data, offset);
+
+            if flags & range::RANGE_EMPTY > 0 {
+                assert!(
+                    flags == range::RANGE_EMPTY,
+                    "empty ranges contain only RANGE_EMPTY flag"
+                );
+
+                return Datum::Range(Range { inner: None });
+            }
+
+            let lower_bound = if flags & range::RANGE_LB_INF > 0 {
+                None
+            } else {
+                Some(DatumNested::extract(data, offset))
+            };
+
+            let lower = RangeBound {
+                inclusive: flags & range::RANGE_LB_INC > 0,
+                bound: lower_bound,
+            };
+
+            let upper_bound = if flags & range::RANGE_UB_INF > 0 {
+                None
+            } else {
+                Some(DatumNested::extract(data, offset))
+            };
+
+            let upper = RangeBound {
+                inclusive: flags & range::RANGE_UB_INC > 0,
+                bound: upper_bound,
+            };
+
+            Datum::Range(Range {
+                inner: Some(RangeInner { lower, upper }),
+            })
+        }
```

Comment:
Nice! This looks fantastic. I think `bitflags` would make this read marginally better, but the encoding itself looks 👌🏽.